Parse XML sitemaps via SAX for option to parse partial/broken documents

jnioche opened the pull request:
Updated #80 - merged with master branch, fixed a few issues, applied formatting, fix headers etc...

Will compare the perfs on some large sitemap files. I'd expect if not a gain in speed, then at least one in memory usage (since we don't hold the whole DOM representation in RAM).

Have closed #80, let's resume the work here.

@matt-deboer sorry it took so long to have another look at your work

jnioche commented:
I ran a benchmark on this branch vs existing DOM parsing. As expected is uses a lot less memory (90MB vs 180MB on my test docs)

[https://www.novinky.cz/sitemap/sitemap_articles_5.xml]
[http://www.foxnews.com/sitemap.xml?idx=26]
[http://push.abs-cbn.com/sitemap/2015/]

kkrugler commented on src/main/java/crawlercommons/sitemaps/AbstractSiteMapSAXHandler.java:
```diff
@@ -0,0 +1,130 @@
+    }
+
+    public void endElement(String uri, String localName, String qName) throws SAXException {
+        if (delegate != null)
```
add code block for consistency

jnioche replied:
Fixed in b9d5128

kkrugler commented on src/main/java/crawlercommons/sitemaps/AbstractSiteMapSAXHandler.java:
```diff
@@ -0,0 +1,130 @@
+ * Provides a base SAX handler for parsing of XML documents representing
+ * sub-classes of AbstractSiteMap.
+ */
+public class AbstractSiteMapSAXHandler extends DefaultHandler {
```
Seems odd to call this class AbstractXXX when it's not abstract.

jnioche replied:
I agree. Would DefaultSiteMapSAXHandler be a better name? or rename SiteMapSAXHandler into  something else and use SiteMapSAXHandler for this one?

jnioche replied:
b9d5128 -> renamed the classes. @kkrugler let me know what you think of it

kkrugler commented on src/main/java/crawlercommons/sitemaps/AbstractSiteMapSAXHandler.java:
```diff
@@ -0,0 +1,130 @@
+    }
+
+    protected String currentElementParent() {
+        return (elementStack.size() < 2) ? null : elementStack.get(1);
```
Shouldn't this be a get(<list size> - 2), not get(1)?

jnioche replied:
do you mean `get(getSize() - 2)` ? Makes sense.

jnioche replied:
actually the existing code is correct. The more recent elements are added to the head of the queue so the element parent will always be in position 1. For some reason I expected them to be put at the end.

kkrugler commented on src/main/java/crawlercommons/sitemaps/SiteMapIndexSAXHandler.java:
```diff
@@ -0,0 +1,104 @@
+        lastMod = null;
+    }
+
+    public void error(SAXParseException e) throws SAXException {
```
Why do we maybe want to add the sitemap if we get an error while parsing? Is this part of the "handle broken sitemaps" logic? If so, a comment would be good.

jnioche replied:
I can only second guess what @matt-deboer intended. It could be to make sure the loc and lastmod fields are reset to null, however a new instance of the handler class would be created for a  new doc anyway. It could help with the "handle broken sitemaps" if the exception occurred between the discovery of the location and the next sitemap or sitemapindex tag.

kkrugler commented on src/main/java/crawlercommons/sitemaps/SiteMapRSSSAXHandler.java:
```diff
@@ -0,0 +1,153 @@
+     * .String, java.lang.String, java.lang.String, org.xml.sax.Attributes)
+     */
+    @Override
+    public void startElement(String uri, String localName, String qName, Attributes attributes) throws SAXException {
```
Doesn't this need to call super.startElement? Otherwise the element stack isn't going to be set up, right?

kkrugler commented on src/main/java/crawlercommons/sitemaps/SiteMapSAXHandler.java:
```diff
@@ -0,0 +1,118 @@
+        sitemap.setType(SitemapType.XML);
+    }
+
+    public void startElement(String uri, String localName, String qName, Attributes attributes) throws SAXException {
```
Doesn't super.startElement() need to be called? Or (confusingly) is this class only used as a delegate from AbstractSiteMapSAXHandler? Something seems odd here.

jnioche replied:
I think this is these classes are used only as delegates.

jnioche commented on src/main/java/crawlercommons/sitemaps/AbstractSiteMapSAXHandler.java:
```diff
@@ -0,0 +1,130 @@
+        return (elementStack.size() < 2) ? null : elementStack.get(1);
+    }
+
+    public AbstractSiteMap getSiteMap() {
```
Overlooked I suppose

jnioche replied:
Fixed in b9d5128

jnioche commented:
The SAX parsers do not work for cases like #29. Will add a test class to the master branch and apply it to this branch as well + fix some of the issues highlighted by @kkrugler.

jnioche commented:
Have identified a possible issue with the locations not being properly handled. Will add a test case to the master branch first.

@kkrugler I have addressed some of the issues you mentioned, will look again at the startElement() logic.

jnioche commented:
Was an issue with locations split over several calls to characters() - as illustrated by d200346

Still need to apply the same logic to XMLIndex, Atom and RSS Handlers and modify tests to cover for these cases.

The latest commit also improves the efficiency of the parsing as URLs were created twice from strings unnecessarily.

sebastian-nagel commented:
I've adapted this PR to master and for testing and evaluation doubled the parser classes (added SiteMapParserSAX), see [this fork](https://github.com/sebastian-nagel/crawler-commons/tree/cc-116-sitemap-sax-parser). After testing both classes on 35 problematic sitemaps and feeds: I've found no regressions, truncated docs are successfully parsed as expected, XML syntax errors do not affect the whole document, but only the tail after the error.

There are two option how to proceed:
1. replace the existing DOM-based parser
2. keep both parsers and let the user decide which one to use (cf. [--use-sax-parser in SiteMapTester](https://github.com/sebastian-nagel/crawler-commons/blob/cc-116-sitemap-sax-parser/src/main/java/crawlercommons/sitemaps/SiteMapTester.java))

kkrugler commented:
Hi @sebastian-nagel - so what would be a reason for keeping the DOM-based parser around? From @jnioche's comments above, SAX should be faster, use less memory, and should be more resilient to errors (would this also take care of #148?).

sebastian-nagel commented:
For the long-term, I principally agree: maintaining two parsers means additional work, and there are few arguments in favor of the DOM-based parser:
- less code, 700 vs. 1000 lines, excluding license headers, ev. easier to read
- optionally expose the DOM or parts of it? For the more complex, derived formats (images #36, videos #35, news) this could be a quick and extensible solution to make the additional information accessible without the need to define huge classes to hold all the information.

> should be more resilient to errors

Right now, the only difference is that the SAX parser is able to extract URLs from the top of the XML document. It does not recover from an error. Would be nice if the parser could continue after trivial errors such as in #148. I don't know whether this is doable based on the existing code. I found one example ([1](https://gist.github.com/chrisguitarguy/1305010), [2](http://blog.viktorpetersson.com/post/115153423414/how-to-parse-and-dump-a-sitemap)) in Python based on [beautifulsoup](https://pypi.python.org/pypi/beautifulsoup4).

> SAX should be faster, use less memory

In addition, it could provide an iterator over SiteMapURLs. Often URLs are filtered or sampled, and there is no need to keep the complete list of 50,000 URLs (or more for sitemap indexes) in memory. 

For the next time, I want to test the SAX-based parser and compare it to the DOM-based one. I'll prepare a larger test set (few 1000 sitemaps), right now I can share a list of 16 million URLs pointing to sitemaps.

jnioche commented:
+1 to having both for the time being.  @sebastian-nagel could you please open a new PR?

sebastian-nagel commented:
Done: #150. I've used the SAX parser to process 500,000 sitemaps, but have to mine the logs for potential regressions.

jnioche commented:
>  I've used the SAX parser to process 500,000 sitemaps

did you monitor the processing time and memory used compared to the DOM based one?

sebastian-nagel commented:
No, it was a mapreduce job, and also only an overlapping set of URLs. For performance measures, it's probably also better to use a sample on disk not URLs. A first comparison of the log files also shows that the Tika version (also changed) has some impact whether formats are properly recognized. I'll sample a subset of URLs (excluding sitemap indexes) and pre-fetch it for measuring performance, and to have a controlled environment with no differences except the parser.

jnioche commented:
Replaced by #150
